homeworks/solutions/nico778/4-param/src/halfedge.js

```diff
@@ -0,0 +1,649 @@
+
+          // point the new halfedge and face to each other
+          boundaryHalfedge.face = f
+          f.halfedge = boundaryHalfedge
+
+          // point the new halfedge and twin to each other
+          boundaryHalfedge.twin = current
+          current.twin = boundaryHalfedge
+
+          current = nextHe
+        } while(current != he)
+
+        // link the cycle of boundary halfedges together
+        const n = boundaryCycle.length
+        for (let j = 0; j < n; j++) {
+          boundaryCycle[j].next = boundaryCycle[(j+n-1)%n]
+          boundaryCycle[j].prev = boundaryCycle[(j+1)%n]
+          hasTwinHalfedge.set(boundaryCycle[j], true)
+          hasTwinHalfedge.set(boundaryCycle[j].twin, true)
+        }
+      }
+    }
+
+    // allocate indices for all elements
+    let index = 0
+    this.vertices.forEach(v => {
+      v.idx = index++
+    })
+    index = 0
+    this.vertsOrig.forEach(v => {
+      v.idx = index++
+    })
+    index = 0
+    this.edges.forEach(e => {
+      e.idx = index++
+    })
+    index = 0
+    this.faces.forEach(f => {
+      f.idx = index++
+    })
+    index = 0
+    this.halfedges.forEach(he => {
+      // HACK: all halfedge now accessible to the original vertices
+      he.vertsOrig = this.vertsOrig
+      he.idx = index++
+    })
+
+  }
+  /**
+   * flatten computes the UV coordinates of the given triangle mesh.
+   *
+   * This implementation reuiqres the mesh contains at least one
+   * boundary loop.
+   *
+   * @param {string} boundaryType 'disk', or 'rect'
+   * @param {string} laplaceWeight 'uniform', or 'cotan'
+   */
+
+  flatten(boundaryType, laplaceWeight) {
+    // TODO: implement the Tutte's barycentric embedding algorithm.
+    //
+    // Hint:
+    // 1. check if the mesh contains a boundary
+    // 2. compute boundary uv coordinates depending on the boundary type
+    // 3. compute matrix depending on the laplacian weight type
+    // 4. solve linear equation and assing computed uv as vertex uv
+    
+    console.log(this.boundaries.length)
+
+		let boundaryVertices = 0
+    
+    this.boundaries[0].vertices(v => {
+      boundaryVertices++
+    })
+		
+    console.log(boundaryVertices)
+
+    let boundaryEdges = []
+    let tempHe = this.boundaries[0].halfedge
+
+    //let boundaryLength = 0
+
+    for(let n = 0; n < boundaryVertices; n++) {
+      //boundaryLength += tempHe.vector().norm()
+      boundaryEdges.push(tempHe)
+      tempHe = tempHe.next 
+    }
+
+    let uCoo = DenseMatrix.zeros(this.vertices.length)
+    let vCoo = DenseMatrix.zeros(this.vertices.length)
+    const radius = 0.5
+
+    switch(boundaryType) {
+      case 'disk':
+        boundaryEdges.forEach((halfedge, i) => {
+          uCoo.set((radius*Math.cos(2*Math.PI*i/boundaryEdges.length) + radius), halfedge.vertex.idx)
+          vCoo.set((radius*Math.sin(2*Math.PI*i/boundaryEdges.length) + radius), halfedge.vertex.idx)
+        })
+
+        break
+      case 'rect':
```

Comment:
Interesting. Looks like everybody is a little bit struggling with how to get a nicer implementation that produces the boundary vertex positions. But overall the current approach works fine, which fulfills the purpose of the homework :)